### PR TITLE
[FLINK-33067] Expose rate limiter config and enable by default

### PR DESCRIPTION
**Pull request by gyfora**
## What is the purpose of the change

There are certain cases currently when the operator can stuck in an infinite no wait retry loop under certain circumstances (see https://github.com/operator-framework/java-operator-sdk/issues/2046 for details).

As a solution we introduce rate limiter configuration with a default that should already safeguard against prod problems that may arise without it.

The PR also fixes and improves the default retry configuration and fixes a bug with exponential backoff.

## Brief change log

  - *Expose rate limiter config and enable by default*
  - *Update default retry config*
  - *Unit Tests*

## Verifying this change

Unit tests, manual verification

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changes to the `CustomResourceDescriptors`: no
  - Core observer or reconciler logic that is regularly executed: yes

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? docs

### CODE REVIEW

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorConfigOptions.java [hunk 1]
```diff
@@ -365,22 +365,45 @@ public static String operatorConfigKey(String key) {
...
+                    .defaultValue(Duration.ofSeconds(15))
+                    .withDescription("Operator rate limiter refresh period for each resource.");
+
+    @Documentation.Section(SECTION_SYSTEM)
+    public static final ConfigOption<Integer> OPERATOR_RATE_LIMITER_LIMIT =
+            operatorConfig("rate-limiter.limit")
+                    .intType()
+                    .defaultValue(5)
+                    .withDescription(
+                            "Max number of reconcile loops triggered in within the rate limiter refresh period for each resource. Setting the limit <= 0 disables the limiter.");
```
Comment (mbalassi) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorConfigOptions.java:405]: nit: `triggered within`

**Review by mbalassi — APPROVED** [review]
Thanks, great. Could you please explain the default behavior in plain English now given the default values you propose? (How many request would we send/resources/sec)

**Comment by gyfora** [comment]
> Thanks, great. Could you please explain the default behavior in plain English now given the default values you propose? (How many request would we send/resources/sec)

In plain English the default is:
At most 5 reconciles for a given resource every 15 seconds. This should be more than enough to not have any visible effect on the normal behaviour (upgrades would go through instantaneously as they need only 2/3 reconciles)

But if it gets stuck in an error loop, it will only do 5 loops, then wait for 15 seconds, rinse and repeat

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorConfigOptions.java [hunk 2]
```diff
@@ -365,22 +365,45 @@ public static String operatorConfigKey(String key) {
                     .withDescription(
                             "Initial interval of automatic reconcile retries on recoverable errors.");
 
+    @Documentation.Section(SECTION_SYSTEM)
+    public static final ConfigOption<Duration> OPERATOR_RETRY_MAX_INTERVAL =
+            operatorConfig("retry.max.interval")
+                    .durationType()
+                    .defaultValue(Duration.ZERO)
+                    .withDescription(
+                            "Max interval of automatic reconcile retries on recoverable errors, if set to <=0, no limit is applied.");
```
Comment (mxm) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorConfigOptions.java:374]: How are recoverable errors defined?

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorConfigOptions.java [hunk 3]
```diff
@@ -365,22 +365,45 @@ public static String operatorConfigKey(String key) {
...
+
     @Documentation.Section(SECTION_SYSTEM)
     public static final ConfigOption<Double> OPERATOR_RETRY_INTERVAL_MULTIPLIER =
             operatorConfig("retry.interval.multiplier")
                     .doubleType()
-                    .defaultValue(2.0)
+                    .defaultValue(1.5)
```
Comment (mxm) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorConfigOptions.java:380]: Why are we changing this value here? It's unrelated to the rate limiting

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorConfigOptions.java [hunk 4]
```diff
@@ -365,22 +365,45 @@ public static String operatorConfigKey(String key) {
...
                     .withDescription(
                             "Interval multiplier of automatic reconcile retries on recoverable errors.");
 
     @Documentation.Section(SECTION_SYSTEM)
     public static final ConfigOption<Integer> OPERATOR_RETRY_MAX_ATTEMPTS =
             operatorConfig("retry.max.attempts")
                     .intType()
-                    .defaultValue(10)
+                    .defaultValue(15)
```
Comment (mxm) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorConfigOptions.java:388]: Same here.

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorConfigOptions.java [hunk 5]
```diff
@@ -365,22 +365,45 @@ public static String operatorConfigKey(String key) {
...
                     .withDescription(
                             "Max attempts of automatic reconcile retries on recoverable errors.");
 
+    @Documentation.Section(SECTION_SYSTEM)
+    public static final ConfigOption<Duration> OPERATOR_RATE_LIMITER_PERIOD =
+            operatorConfig("rate-limiter.refresh-period")
+                    .durationType()
+                    .defaultValue(Duration.ofSeconds(15))
+                    .withDescription("Operator rate limiter refresh period for each resource.");
```
Comment (mxm) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorConfigOptions.java:397]:
```suggestion
            operatorConfig("rate-limiter.period")
                    .durationType()
                    .defaultValue(Duration.ofSeconds(15))
                    .withDescription("Operator rate limiter period for each resource.");
```

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkOperatorConfiguration.java [hunk 6]
```diff
@@ -220,8 +227,14 @@ private static GenericRetry getRetryConfig(Configuration conf) {
                                 .toMillis())
                 .setIntervalMultiplier(
                         conf.getDouble(
-                                KubernetesOperatorConfigOptions
-                                        .OPERATOR_RETRY_INTERVAL_MULTIPLIER));
+                                KubernetesOperatorConfigOptions.OPERATOR_RETRY_INTERVAL_MULTIPLIER))
+                .setMaxInterval(maxInterval > 0 ? maxInterval : Long.MAX_VALUE);
```
Comment (mxm) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkOperatorConfiguration.java:231]: The default value `Long.MAX_VALUE` means, retry as often as you like? Any value than that means wait at least that amount of time in milliseconds?

Comment (mxm) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkOperatorConfiguration.java:231]: It's a bit tricky because the JOSDK treats Long.MAX_VALUE as unset...

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorConfigOptions.java (hunk 5 above)
Comment (mxm) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorConfigOptions.java:397]: I guess it means, that we refresh 15 seconds after hitting the limit. Current name makes sense then.

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorConfigOptions.java (hunk 2 above)
Comment (mxm) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorConfigOptions.java:374]: AFAIK the current default wasn't modified by adding this configuration option, correct?

**Review by mxm — APPROVED** [review]
Generally, looks good. Just some clarifications above.

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkOperatorConfiguration.java (hunk 6 above)
Comment (gyfora) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkOperatorConfiguration.java:231]: I don't really see this MAX_VALUE behaviour that you mention, in any case I changed this to -1 which is the right way to set infinite

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorConfigOptions.java (hunk 2 above)
Comment (gyfora) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorConfigOptions.java:374]: Previously it was infinite, and a recent change actually introduced a 15sec limit accidentally. This is a correction of that. So with this, no change since the last release :)

Comment (gyfora) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorConfigOptions.java:374]: As for the "recoverable error", I just lifted the wording from the other config. It basically means any error that is not handled by the operator logic directly but is "thrown" and therefore should be retried :)

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorConfigOptions.java (hunk 3 above)
Comment (gyfora) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorConfigOptions.java:380]: Based on some testing I felt that the current 2 multiplier for the exponential backoff was a bit too aggressive in some cases. So 1.5 with a little longer max retries felt more appropriate. If you feel we shouldn't change this I remove it from the PR

**Comment by gyfora** [comment]
@mxm I have updated the PR based on the comments, also simplified wording for the retry related configs. PTAL

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkOperatorConfiguration.java (hunk 6 above)
Comment (mxm) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkOperatorConfiguration.java]: See here for default `UNSET_VALUE`: https://github.com/operator-framework/java-operator-sdk/blob/a3fb3ad9ef84ad60e456ac2b3714b070b37e055b/operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/retry/GenericRetry.java#L103 

which is set to `Long.MAX_VALUE`: https://github.com/operator-framework/java-operator-sdk/blob/a3fb3ad9ef84ad60e456ac2b3714b070b37e055b/operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/retry/GradualRetry.java#L19

Comment (mxm) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkOperatorConfiguration.java]: If set to `Long.MAX_VALUE`, this default will be applied: https://github.com/operator-framework/java-operator-sdk/blob/a3fb3ad9ef84ad60e456ac2b3714b070b37e055b/operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/retry/GradualRetry.java#L16

Comment (gyfora) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkOperatorConfiguration.java]: Oh I see what you meant, this wouldn't affect us as we are not using the `@GradualRetry` annotation, that is specific to that. In any case I removed this to be less confusing
